File: models/network_planning.py
"""
Network Planning Schema v1.0

Python dataclass models for network planning state.
These models mirror the TypeScript schema and provide
conversion to/from dict for JSON serialization.
"""

from dataclasses import dataclass, field, asdict
from typing import List, Optional, Dict, Any
from datetime import datetime

SCHEMA_VERSION = "1.0"


@dataclass
class AddressPrefix:
    """VPC address prefix."""
    id: str
    name: str
    cidr: str
    zone: str
    is_default: bool = False


@dataclass
class VpcPlan:
    """VPC planning configuration."""
    id: str
    name: str
    region: str
    address_prefix_mode: str = "manual"
    address_prefixes: List[AddressPrefix] = field(default_factory=list)
    resource_group_id: Optional[str] = None
    tags: Dict[str, str] = field(default_factory=dict)
    notes: str = ""
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())
    updated_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())

    def __post_init__(self):
        if self.address_prefix_mode not in ("manual", "auto"):
            raise ValueError("address_prefix_mode must be 'manual' or 'auto'")
        # Convert dict address_prefixes to AddressPrefix objects
        if self.address_prefixes and isinstance(self.address_prefixes[0], dict):
            self.address_prefixes = [
                AddressPrefix(**ap) for ap in self.address_prefixes
            ]


@dataclass
class SubnetPlan:
    """Subnet planning configuration."""
    id: str
    name: str
    vpc_id: str
    zone: str
    cidr: str
    purpose: str = ""
    source_network: Optional[str] = None
    public_gateway: bool = False
    public_gateway_id: Optional[str] = None
    acl_id: Optional[str] = None
    route_table_id: Optional[str] = None
    ipv4_cidr_count: Optional[int] = None
    tags: Dict[str, str] = field(default_factory=dict)
    notes: str = ""
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())
    updated_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())


@dataclass
class SecurityRule:
    """Security group rule."""
    id: str
    direction: str
    protocol: str
    port_min: Optional[int] = None
    port_max: Optional[int] = None
    source: Optional[str] = None
    destination: Optional[str] = None
    description: str = ""

    def __post_init__(self):
        if self.direction not in ("inbound", "outbound"):
            raise ValueError("direction must be 'inbound' or 'outbound'")
        if self.protocol not in ("tcp", "udp", "icmp", "all"):
            raise ValueError("protocol must be 'tcp', 'udp', 'icmp', or 'all'")


@dataclass
class SecurityGroupPlan:
    """Security group planning configuration."""
    id: str
    name: str
    vpc_id: str
    purpose: str = ""
    rules: List[SecurityRule] = field(default_factory=list)
    tags: Dict[str, str] = field(default_factory=dict)
    notes: str = ""
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())
    updated_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())

    def __post_init__(self):
        # Convert dict rules to SecurityRule objects
        if self.rules and isinstance(self.rules[0], dict):
            self.rules = [SecurityRule(**rule) for rule in self.rules]


@dataclass
class StorageProfilePlan:
    """Storage profile planning configuration."""
    id: str
    name: str
    tier: str
    iops_intent: str = ""
    notes: str = ""
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())
    updated_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())


@dataclass
class WavePlan:
    """Migration wave planning configuration."""
    id: str
    name: str
    owner: str = ""
    target_window: str = ""
    priority: str = "medium"
    notes: str = ""
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())
    updated_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())


@dataclass
class NetworkComponentPlan:
    """Network component planning configuration."""
    id: str
    name: str
    type: str
    vpc_id: str
    subnet_id: Optional[str] = None
    attachment: str = ""
    config: Dict[str, Any] = field(default_factory=dict)
    tags: Dict[str, str] = field(default_factory=dict)
    notes: str = ""
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())
    updated_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())


@dataclass
class SecondaryNicAssignment:
    """Secondary NIC assignment."""
    id: str
    subnet_id: str
    security_group_id: str
    order: int
    source_network: Optional[str] = None


@dataclass
class VmNetworkAssignment:
    """VM network assignment."""
    vm_key: str
    vm_name: str
    primary_subnet_id: str
    primary_security_group_id: str
    secondary_nics: List[SecondaryNicAssignment] = field(default_factory=list)
    storage_profile_id: Optional[str] = None
    wave_id: Optional[str] = None
    excluded: bool = False
    exclusion_reason: Optional[str] = None

    def __post_init__(self):
        # Convert dict secondary_nics to SecondaryNicAssignment objects
        if self.secondary_nics and isinstance(self.secondary_nics[0], dict):
            self.secondary_nics = [
                SecondaryNicAssignment(**nic) for nic in self.secondary_nics
            ]


@dataclass
class PlanningMetadata:
    """Planning metadata."""
    project_name: str = ""
    target_region: str = "us-south"
    target_zone: str = "us-south-1"
    deployment_target: str = "plain_cli"
    created_by: Optional[str] = None
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())
    updated_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())
    rvtools_filename: Optional[str] = None
    rvtools_uploaded_at: Optional[str] = None


@dataclass
class NetworkPlanningState:
    """Root network planning state object."""
    version: str = SCHEMA_VERSION
    vpcs: List[VpcPlan] = field(default_factory=list)
    subnets: List[SubnetPlan] = field(default_factory=list)
    security_groups: List[SecurityGroupPlan] = field(default_factory=list)
    storage_profiles: List[StorageProfilePlan] = field(default_factory=list)
    waves: List[WavePlan] = field(default_factory=list)
    network_components: List[NetworkComponentPlan] = field(default_factory=list)
    vm_assignments: List[VmNetworkAssignment] = field(default_factory=list)
    metadata: PlanningMetadata = field(default_factory=PlanningMetadata)

    def __post_init__(self):
        # Convert dict objects to dataclass instances
        if self.vpcs and isinstance(self.vpcs[0], dict):
            self.vpcs = [VpcPlan(**vpc) for vpc in self.vpcs]
        if self.subnets and isinstance(self.subnets[0], dict):
            self.subnets = [SubnetPlan(**subnet) for subnet in self.subnets]
        if self.security_groups and isinstance(self.security_groups[0], dict):
            self.security_groups = [
                SecurityGroupPlan(**sg) for sg in self.security_groups
            ]
        if self.storage_profiles and isinstance(self.storage_profiles[0], dict):
            self.storage_profiles = [
                StorageProfilePlan(**sp) for sp in self.storage_profiles
            ]
        if self.waves and isinstance(self.waves[0], dict):
            self.waves = [WavePlan(**wave) for wave in self.waves]
        if self.network_components and isinstance(self.network_components[0], dict):
            self.network_components = [
                NetworkComponentPlan(**nc) for nc in self.network_components
            ]
        if self.vm_assignments and isinstance(self.vm_assignments[0], dict):
            self.vm_assignments = [
                VmNetworkAssignment(**va) for va in self.vm_assignments
            ]
        if isinstance(self.metadata, dict):
            self.metadata = PlanningMetadata(**self.metadata)


# Conversion helpers
def to_dict(obj: Any) -> Dict[str, Any]:
    """Convert dataclass to dict recursively."""
    if hasattr(obj, '__dataclass_fields__'):
        return asdict(obj)
    return obj


def from_dict(data: Dict[str, Any]) -> NetworkPlanningState:
    """Convert dict to NetworkPlanningState."""
    return NetworkPlanningState(**data)

# Made with Bob

File: prototype/carbon-ui/types/network-planning.ts
/**
 * Network Planning Schema v1.0
 *
 * This schema defines the structure for network planning state
 * that bridges Carbon UI planning and Terraform generation.
 */

export const SCHEMA_VERSION = '1.0';

// Core Types
export type NetworkPlanningState = {
  version: string;
  vpcs: VpcBucket[];
  subnets: SubnetBucket[];
  securityGroups: SecurityBucket[];
  storageProfiles: StorageBucket[];
  waves: WaveBucket[];
  networkComponents: NetworkComponent[];
  vmAssignments: VmNetworkAssignment[];
  metadata: PlanningMetadata;
};

export type VpcBucket = {
  id: string;
  name: string;
  label: string;
  region: string;
  addressPrefixMode: 'manual' | 'auto';
  addressPrefixes: AddressPrefix[];
  resourceGroupId?: string;
  tags: Record<string, string>;
  notes: string;
  createdAt: string;
  updatedAt: string;
};

export type AddressPrefix = {
  id: string;
  name: string;
  cidr: string;
  zone: string;
  isDefault: boolean;
};

export type SubnetBucket = {
  id: string;
  name: string;
  label: string;
  vpcId: string;
  zone: string;
  cidr: string;
  purpose: string;
  sourceNetwork?: string;
  publicGateway: boolean;
  publicGatewayId?: string;
  aclId?: string;
  routeTableId?: string;
  ipv4CidrCount?: number;
  tags: Record<string, string>;
  notes: string;
  createdAt: string;
  updatedAt: string;
};

export type SecurityBucket = {
  id: string;
  name: string;
  label: string;
  vpcId: string;
  purpose: string;
  rules: SecurityRule[];
  tags: Record<string, string>;
  notes: string;
  createdAt: string;
  updatedAt: string;
};

export type SecurityRule = {
  id: string;
  direction: 'inbound' | 'outbound';
  protocol: 'tcp' | 'udp' | 'icmp' | 'all';
  portMin?: number;
  portMax?: number;
  source?: string;
  destination?: string;
  description: string;
};

export type StorageBucket = {
  id: string;
  name: string;
  label: string;
  tier: string;
  iopsIntent: string;
  notes: string;
  createdAt: string;
  updatedAt: string;
};

export type WaveBucket = {
  id: string;
  name: string;
  owner: string;
  targetWindow: string;
  priority: 'high' | 'medium' | 'low';
  notes: string;
  createdAt: string;
  updatedAt: string;
};

export type NetworkComponent = {
  id: string;
  name: string;
  label: string;
  type: NetworkComponentType;
  vpcId: string;
  subnetId?: string;
  attachment: string;
  config: Record<string, any>;
  tags: Record<string, string>;
  notes: string;
  createdAt: string;
  updatedAt: string;
};

export type NetworkComponentType =
  | 'public_gateway'
  | 'vpn_gateway'
  | 'load_balancer'
  | 'vpe_gateway'
  | 'floating_ip'
  | 'route_table'
  | 'acl';

export type VmNetworkAssignment = {
  vmKey: string;
  vmName: string;
  primarySubnetId: string;
  primarySecurityGroupId: string;
  secondaryNics: SecondaryNic[];
  storageProfileId?: string;
  waveId?: string;
  excluded: boolean;
  exclusionReason?: string;
  // Compute specifications for VSI generation
  cpuCount?: number;
  memoryGb?: number;
  ibmProfile?: string;
  overrideProfile?: string;
  overrideProfileReason?: string;
  // Boot disk specifications
  bootDiskGb?: number;
  // Custom image reference
  customImageId?: string;
  guestOs?: string;
};

export type SecondaryNic = {
  id: string;
  subnetId: string;
  securityGroupId: string;
  order: number;
  sourceNetwork?: string;
};

export type PlanningMetadata = {
  projectName: string;
  targetRegion: string;
  targetZone: string;
  deploymentTarget: 'plain_cli' | 'schematics';
  sshPublicKey?: string;
  sshKeyName?: string;
  resourceGroupId?: string;
  backendType: 'local' | 's3' | 'cos';
  createdBy?: string;
  createdAt: string;
  updatedAt: string;
  rvtoolsFilename?: string;
  rvtoolsUploadedAt?: string;
};

// Helper Functions
export function createEmptyNetworkPlan(): NetworkPlanningState {
  return {
    version: SCHEMA_VERSION,
    vpcs: [],
    subnets: [],
    securityGroups: [],
    storageProfiles: [],
    waves: [],
    networkComponents: [],
    vmAssignments: [],
    metadata: {
      projectName: '',
      targetRegion: 'us-south',
      targetZone: 'us-south-1',
      deploymentTarget: 'plain_cli',
      backendType: 'local',
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    },
  };
}

export function createInitialNetworkPlan(projectName: string): NetworkPlanningState {
  const now = new Date().toISOString();
  const vpcId = crypto.randomUUID();
  const subnetId = crypto.randomUUID();
  const sgId = crypto.randomUUID();

  return {
    version: SCHEMA_VERSION,
    vpcs: [
      {
        id: vpcId,
        name: 'migration-vpc',
        label: 'Migration VPC',
        region: 'us-south',
        addressPrefixMode: 'manual',
        addressPrefixes: [
          {
            id: crypto.randomUUID(),
            name: 'zone-1-prefix',
            cidr: '10.240.0.0/16',
            zone: 'us-south-1',
            isDefault: true,
          },
        ],
        tags: {},
        notes: '',
        createdAt: now,
        updatedAt: now,
      },
    ],
    subnets: [
      {
        id: subnetId,
        name: 'default-subnet',
        label: 'Default Subnet',
        vpcId,
        zone: 'us-south-1',
        cidr: '10.240.10.0/24',
        purpose: 'general',
        publicGateway: false,
        tags: {},
        notes: '',
        createdAt: now,
        updatedAt: now,
      },
    ],
    securityGroups: [
      {
        id: sgId,
        name: 'default-sg',
        label: 'Default Security Group',
        vpcId,
        purpose: 'general',
        rules: [
          {
            id: crypto.randomUUID(),
            direction: 'outbound',
            protocol: 'all',
            description: 'Allow all outbound',
          },
        ],
        tags: {},
        notes: '',
        createdAt: now,
        updatedAt: now,
      },
    ],
    storageProfiles: [],
    waves: [],
    networkComponents: [],
    vmAssignments: [],
    metadata: {
      projectName,
      targetRegion: 'us-south',
      targetZone: 'us-south-1',
      deploymentTarget: 'plain_cli',
      backendType: 'local',
      createdAt: now,
      updatedAt: now,
    },
  };
}

// ─── UI-level types (moved from app/page.tsx) ───────────────────────────────

export type EstateSummary = {
  in_scope: number;
  excluded: number;
  monthly: number;
  savings: number;
  blocked: number;
  review: number;
};

export type WorkbookSummary = {
  filename: string;
  estate_summary: EstateSummary;
  overview_blockers: Record<string, number>;
  readiness_counts: Record<string, Record<string, number>>;
  assessment_quality: Record<string, string | number>;
  readiness_rows: Array<Record<string, unknown>>;
  assignment_rows?: Array<Record<string, unknown>>;
};

/** UI-level flattened view of a VM row — distinct from VmNetworkAssignment (API shape). */
export type AssignmentVm = {
  id: string;
  name: string;
  image: string;
  imageReasons: string;
  originalSpecs?: string;
  migration: string;
  migrationReasons: string;
  memory: string;
  memoryReasons: string;
  networkReadiness: string;
  networkReasons: string;
  profile: string;
  overrideProfile: string;
  overrideProfileReason?: string;
  storageTier: string;
  overrideStorageTier: string;
  overrideStorageTierReason?: string;
  network: string;
  subnet: string;
  securityGroup: string;
  power: string;
  guestOs?: string;
  sourceIp?: string;
  datacenter?: string;
  cluster?: string;
  host?: string;
  diskCount?: string;
  dataDiskCount?: string;
  totalStorageGb?: string;
  firmware?: string;
  bootDiskGb?: string;
  configuredMemoryMib?: string;
  activeMemoryMib?: string;
  consumedMemoryMib?: string;
  balloonedMemoryMib?: string;
  swappedMemoryMib?: string;
  memoryReservationMib?: string;
  memoryLimitMib?: string;
  memoryHotAdd?: string;
  sizingMemoryMib?: string;
  memorySizingBasis?: string;
  diskDetails?: unknown[];
  partitionDetails?: unknown[];
  partitionCount?: string;
  unmatchedPartitionCount?: string;
  networkDetails?: unknown[];
  readinessFindings?: unknown[];
  networkReadinessFindings?: unknown[];
  computeMonthly?: string;
  storageMonthly?: string;
  monthlyCost?: string;
  baselineCostMonthly?: string;
  savingsMonthly?: string;
  pricingSource?: string;
  pricingConfidence?: string;
  pricingLastUpdated?: string;
  pricingStatus?: string;
  profileHourly?: string;
  owner: string;
  application: string;
  wave: string;
  cutoverGroup: string;
  priority: string;
  dependencyGroup: string;
  excluded?: boolean;
  exclusionReason?: string;
  storageLabel?: string;
};

export type SavedProject = {
  id: string;
  name: string;
  description?: string;
  created_at?: string;
  updated_at?: string;
};

export type ResourceState = {
  vpcs: VpcBucket[];
  subnets: SubnetBucket[];
  securityGroups: SecurityBucket[];
  storageProfiles: StorageBucket[];
  waves: WaveBucket[];
  networkComponents: NetworkComponent[];
};

export type SuggestionConfidence = 'High' | 'Medium' | 'Low';

export type SuggestionAuditEntry = {
  id: string;
  vmId: string;
  vmName: string;
  field: 'subnet' | 'securityGroup' | 'storage' | 'wave';
  oldValue: string;
  newValue: string;
  confidence: SuggestionConfidence;
  reason: string;
  evidence: string[];
  appliedAt: string;
  revertedAt?: string;
};

export type SavedProjectState = {
  planning_state_json?: {
    carbon_summary?: WorkbookSummary;
    carbon_assignment_rows?: AssignmentVm[];
    carbon_resources?: ResourceState;
    carbon_remediation_tracker?: RemediationTracker;
    carbon_image_import_status?: ImageImportStatus;
    carbon_suggestion_audit?: SuggestionAuditEntry[];
    remediation_tracker?: Record<string, {
      status?: string;
      owner?: string;
      due_date?: string;
      dueDate?: string;
      notes?: string;
    }>;
    image_import_status?: Record<string, {
      target_catalog_id?: string;
      targetCatalogId?: string;
      import_status?: string;
      importStatus?: string;
      estimated_import_time?: string;
      estimatedImportTime?: string;
      notes?: string;
    }>;
    suggestion_audit?: Array<{
      id?: string;
      vm_id?: string;
      vmId?: string;
      vm_name?: string;
      vmName?: string;
      field?: string;
      old_value?: string;
      oldValue?: string;
      new_value?: string;
      newValue?: string;
      confidence?: string;
      reason?: string;
      evidence?: string[];
      applied_at?: string;
      appliedAt?: string;
      reverted_at?: string | null;
      revertedAt?: string | null;
    }>;
    metadata?: Record<string, string>;
  };
};

export type AssignmentMode = 'network' | 'security' | 'storage' | 'wave';

export type RemediationStatus = 'Open' | 'In Progress' | 'Resolved' | 'Deferred';

export type RemediationTrackerEntry = {
  status: RemediationStatus;
  owner: string;
  dueDate: string;
  notes: string;
  vmKey?: string;
  vm_key?: string;
  blockerType?: string;
  blocker_type?: string;
  blockerDescription?: string;
  blocker_description?: string;
  description?: string;
  type?: string;
};

export type RemediationTracker = Record<string, RemediationTrackerEntry>;

export type RemediationBacklogItem = {
  blockerId: string;
  vmKey: string;
  vmName: string;
  owner: string;
  blockerType: string;
  blockerDescription: string;
  status: RemediationStatus;
  dueDate: string;
  notes: string;
};

export type ImageImportStatusValue =
  | ''
  | 'Pending'
  | 'Scheduled'
  | 'Imported'
  | 'Failed'
  | 'Review';

export type ImageImportEntry = {
  targetCatalogId: string;
  importStatus: ImageImportStatusValue;
  estimatedImportTime: string;
  notes: string;
};

export type ImageImportStatus = Record<string, ImageImportEntry>;

export type ImageImportRow = {
  sourceImage: string;
  vmCount: number;
  owners: string;
  targetCatalogId: string;
  importStatus: ImageImportStatusValue;
  estimatedImportTime: string;
  notes: string;
};

export type CutoverStatus = 'Ready' | 'Review' | 'Blocked';

export type CutoverReadinessRow = {
  vmName: string;
  wave: string;
  cutoverGroup: string;
  owner: string;
  application: string;
  cutoverStatus: CutoverStatus;
  blockerCategory: string;
  blockerReason: string;
  recommendedNextAction: string;
};

export type CutoverSummaryRow = {
  group: string;
  vms: number;
  ready: number;
  review: number;
  blocked: number;
  missingPlanning: number;
  unresolvedRemediation: number;
  imagePending: number;
};

export type Workflow =
  | 'overview'
  | 'intake'
  | 'assignment'
  | 'overrides'
  | 'remediation'
  | 'imageImport'
  | 'migrationOps'
  | 'network'
  | 'security'
  | 'storage'
  | 'waves'
  | 'export';

// Made with Bob
